### ISSUE TITLE
Spurious argument parsing in gfpgan_tools.py

### ISSUE THREAD

**Issue by cmsj** [invoke-ai/InvokeAI#530]
I'm writing some code that uses `ldm.generate` and it's working fine for generating images. I wanted to add upscaler support to my tool, with the new release of this repo.

I added `upscale` and `gfpgan_strength` arguments to my `prompt2image()` call and now I get a usage error after the image has been generated:

```
/Users/administrator/hacking/stable-diffusion/ldm/modules/embedding_manager.py:153: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659484612588/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  placeholder_idx = torch.where(
Generating:   0%|                                                                                                                                                                        | 0/1 [00:00<?, ?it/s]Global seed set to 42
PLMS Sampler: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [02:33<00:00,  3.08s/it]
Generating: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [02:39<00:00, 159.79s/it]
usage: dreambot_backend.py [-h] [--laion400m] [--from_file INFILE] [-n ITERATIONS] [-F] [-g] [-A SAMPLER_NAME] [--outdir OUTDIR] [--seamless] [--embedding_path EMBEDDING_PATH] [--prompt_as_dir]
                           [--gfpgan_bg_upsampler GFPGAN_BG_UPSAMPLER] [--gfpgan_bg_tile GFPGAN_BG_TILE] [--gfpgan_model_path GFPGAN_MODEL_PATH] [--gfpgan_dir GFPGAN_DIR] [--web] [--host HOST]
                           [--port PORT] [--weights WEIGHTS] [--device DEVICE] [--model MODEL] [--config CONFIG]
```

This is coming from [here](https://github.com/lstein/stable-diffusion/blob/main/ldm/gfpgan/gfpgan_tools.py#L10) because it's unconditionally parsing command line arguments even though it's been called as part of `ldm.generate`'s API.

Unless I'm missing something, this makes it impossible to use the upscaling feature outside of `dream.py`.

**Comment by lstein** [comment]
Yes, you are so right. I was looking at this just this morning and thinking that gfpgan's way of getting the options could be a problem down the road. And bingo!

I'll get this fixed sometime tomorrow if @blessedcoolant doesn't beat me to it.

**Comment by blessedcoolant** [comment]
Seems like that file was changed since my implementation. I had a look at it just now and it is calling the boot arguments from the arg parser which makes sense because they are never passed down to `ldm.generate`. All that `ldm.generate` controls is the activation and strength of the restoration. Nothing else.

And that makes sense too because you are basically initializing the dir and path at boot -- which is when you want it to happen. There's no reason to pass that info down to `Generate` unless you are planning on changing these paths mid prompt usage which is almost never the case.

I would like to take a look at your `dreambot_backend.py` file implementation first and see where exactly things are going wrong before I suggest changes to this.

**Comment by cmsj** [comment]
@blessedcoolant sure, the relevant portion is [here](https://github.com/cmsj/dreambot/blob/main/dreambot_backend.py#L110).

tl;dr that function is running on its own thread and it just waits for new incoming prompts to appear on a queue, parses them for arguments, generates the requested image and puts the result on a queue.

FWIW I would think that because Generate is responsible for calling into gfpgan, it should take the 4 extra arguments required. I'd be happy to work up a PR for that, but I figured it was worth starting with an Issue first :)

**Comment by blessedcoolant** [comment]
Yes. Unlike the Dream Script where `dream.py` processes the boot, when you directly load the API, there is nothing that is initializing GFGPAN. The same would be the case for any boot argument that is not passed on to Generate. Feel like we need to streamline this a bit more so the API is fully independent of the Dream script.

Feel free to go ahead and make the PR. I'll have a look when it's up.

**Comment by cmsj** [comment]
If you've already got something, go ahead. Right now all I have is a really ugly bodge job I did locally to get it working at all, but it's using awful global variables :)

**Comment by blessedcoolant** [comment]
Yeah. I like the idea. Like you said, we can't keep bloating up generate or prompt2png's argument list. Maybe the ideal choice is to do a subclass. Not just for this but for almost any other functionality that is NOT dependent on the SD image generation code. This way, anyone who wants to create their own script on top of the API will be able to load and unload modules as per will.

My current fix was just doing to be adding them as args to the generate class which solves the issue but I am a not a fan of the idea because of argument overload in the long run.

**Comment by cmsj** [comment]
>  `p.process(img_path)`

Just curious, and obviously that was just a quick example of how the post processing could be handled, but would this actually be passing around `Image` objects rather than files on disk?

**Comment by blessedcoolant** [comment]
@cmsj #547 -- Something like this should solve your problem.

**Comment by cmsj** [comment]
Nice! Thanks @blessedcoolant

**Comment by psychedelicious** [comment]
The root issue was resolved by the major Restoration class refactor.
